File: python/src/chaos_atlas/maps/logistic.py
"""Logistic map: x -> r x (1 - x).

The canonical route to chaos by period doubling. Chaotic for most r above the
Feigenbaum point r_inf ~= 3.5699.
"""

from __future__ import annotations

import numpy as np
from numpy.typing import NDArray

__all__ = [
    "iterate",
    "trajectory",
    "cobweb",
    "bifurcation",
    "lyapunov",
]


def iterate(x: float, r: float = 3.9) -> float:
    """Advance a single state by one step of the logistic map."""
    return r * x * (1.0 - x)


def trajectory(r: float = 3.9, x0: float = 0.5, iterations: int = 100) -> NDArray[np.float64]:
    """Return the orbit of ``x0``, including ``x0`` itself.

    The returned array has length ``iterations``.
    """
    if iterations < 0:
        raise ValueError("iterations must be non-negative")

    out = np.empty(iterations, dtype=np.float64)
    x = float(x0)
    for i in range(iterations):
        out[i] = x
        x = r * x * (1.0 - x)
    return out


def cobweb(r: float = 3.9, x0: float = 0.5, iterations: int = 100) -> NDArray[np.float64]:
    """Return ``(iterations, 2)`` pairs ``(x_n, x_{n+1})`` for a cobweb plot."""
    if iterations < 0:
        raise ValueError("iterations must be non-negative")

    out = np.empty((iterations, 2), dtype=np.float64)
    x = float(x0)
    for i in range(iterations):
        y = r * x * (1.0 - x)
        out[i, 0] = x
        out[i, 1] = y
        x = y
    return out


def bifurcation(
    r_min: float = 2.5,
    r_max: float = 4.0,
    r_steps: int = 500,
    transient: int = 100,
    iterations: int = 100,
    x0: float = 0.5,
) -> tuple[NDArray[np.float64], NDArray[np.float64]]:
    """Sample the attractor over a range of ``r``.

    Returns ``(r_values, x_values)``, both flat arrays of the same length, so
    they can be handed straight to a scatter plot.

    The transient is discarded per ``r`` so only the attractor is sampled.
    """
    if r_steps < 1:
        raise ValueError("r_steps must be >= 1")

    r_grid = np.linspace(r_min, r_max, r_steps)
    # Vectorised across r: every column is an independent orbit.
    x = np.full(r_steps, float(x0), dtype=np.float64)
    for _ in range(transient):
        x = r_grid * x * (1.0 - x)

    xs = np.empty((iterations, r_steps), dtype=np.float64)
    for i in range(iterations):
        x = r_grid * x * (1.0 - x)
        xs[i] = x

    r_out = np.tile(r_grid, iterations)
    return r_out, xs.ravel()


def lyapunov(
    r: float = 3.9,
    x0: float = 0.5,
    iterations: int = 1000,
    transient: int = 100,
) -> float:
    """Largest Lyapunov exponent, in nats per iteration.

    Uses the exact derivative f'(x) = r (1 - 2x). Positive values indicate
    sensitive dependence on initial conditions.

    Beware preperiodic initial conditions. The exponent describes the orbit you
    actually start, and a few special ``x0`` values fall onto a fixed point
    instead of the attractor. The default ``x0 = 0.5`` is one of them at r = 4:
    it maps to 1.0, then to 0, and stays there, so the result is log 4 rather
    than the log 2 of the chaotic attractor. Pass a generic ``x0`` such as 0.2
    when you want the attractor's exponent::

        logistic.lyapunov(r=4.0)           # 1.3863 = log 4, the orbit at x = 0
        logistic.lyapunov(r=4.0, x0=0.2)   # 0.6931 = log 2, the attractor
    """
    if iterations < 1:
        raise ValueError("iterations must be >= 1")

    x = float(x0)
    for _ in range(transient):
        x = r * x * (1.0 - x)

    total = 0.0
    counted = 0
    for _ in range(iterations):
        derivative = abs(r * (1.0 - 2.0 * x))
        # The orbit can land exactly on the critical point x = 1/2, where the
        # derivative vanishes and log diverges. Skip that sample rather than
        # returning -inf for the whole exponent.
        if derivative > 0.0:
            total += np.log(derivative)
            counted += 1
        x = r * x * (1.0 - x)

    if counted == 0:
        return float("-inf")
    return total / counted

File: lib/maps/logistic.ts
// src/lib/maps/logistic.ts
export interface LogisticPoint {
  x: number;
  y: number;
}

/**
 * Calculate a single iteration of the logistic map
 * @param x Current x value
 * @param r Parameter r (default: 3.9)
 * @returns New x value after one iteration
 */
export function calculateLogisticIteration(
  x: number,
  r: number = 3.9
): number {
  return r * x * (1 - x);
}

/**
 * Calculate the logistic map for a given number of iterations
 * @param r Parameter r (default: 3.9)
 * @param x0 Initial x value (default: 0.5)
 * @param iterations Number of iterations (default: 100)
 * @returns Array of x values representing the logistic map trajectory
 */
export function calculateLogisticMap(
  r: number = 3.9,
  x0: number = 0.5,
  iterations: number = 100
): number[] {
  const points: number[] = [];
  let x = x0;
  
  for (let i = 0; i < iterations; i++) {
    points.push(x);
    x = calculateLogisticIteration(x, r);
  }
  
  return points;
}

/**
 * Calculate the logistic map cobweb plot
 * @param r Parameter r (default: 3.9)
 * @param x0 Initial x value (default: 0.5)
 * @param iterations Number of iterations (default: 100)
 * @returns Array of points representing the cobweb plot
 */
export function calculateLogisticCobweb(
  r: number = 3.9,
  x0: number = 0.5,
  iterations: number = 100
): LogisticPoint[] {
  const points: LogisticPoint[] = [];
  let x = x0;
  
  for (let i = 0; i < iterations; i++) {
    const y = calculateLogisticIteration(x, r);
    points.push({ x, y });
    x = y;
  }
  
  return points;
}

/**
 * Calculate the bifurcation diagram for the logistic map
 * @param rMin Minimum r value (default: 2.5)
 * @param rMax Maximum r value (default: 4.0)
 * @param rSteps Number of r values to calculate (default: 500)
 * @param transient Number of iterations to discard (default: 100)
 * @param iterations Number of iterations to keep (default: 100)
 * @returns Array of points representing the bifurcation diagram
 */
export function calculateLogisticBifurcation(
  rMin: number = 2.5,
  rMax: number = 4.0,
  rSteps: number = 500,
  transient: number = 100,
  iterations: number = 100
): LogisticPoint[] {
  const points: LogisticPoint[] = [];
  const rStep = (rMax - rMin) / rSteps;
  
  for (let rIndex = 0; rIndex <= rSteps; rIndex++) {
    const r = rMin + rIndex * rStep;
    // x0 = 0.5 degenerates to the single value {0} at r = 4 (0.5 -> 1 -> 0
    // -> 0 -> ...), which would draw a misleadingly sparse column in the
    // bifurcation diagram. Use a generic seed instead.
    let x = 0.2;

    // Discard transient
    for (let i = 0; i < transient; i++) {
      x = calculateLogisticIteration(x, r);
    }
    
    // Collect attractor points
    for (let i = 0; i < iterations; i++) {
      x = calculateLogisticIteration(x, r);
      points.push({ x: r, y: x });
    }
  }
  
  return points;
}

/**
 * Calculate the Lyapunov exponent for the logistic map
 * @param r Parameter r (default: 3.9)
 * @param x0 Initial x value (default: 0.5)
 * @param iterations Number of iterations (default: 1000)
 * @returns Approximate Lyapunov exponent
 */
export function calculateLogisticLyapunovExponent(
  r: number = 3.9,
  x0: number = 0.5,
  iterations: number = 1000
): number {
  const transient = 100;
  let x = x0;
  
  // Run the transient iterations
  for (let i = 0; i < transient; i++) {
    x = calculateLogisticIteration(x, r);
  }
  
  // Calculate the Lyapunov exponent
  let sum = 0;
  let counted = 0;
  for (let i = 0; i < iterations; i++) {
    // The derivative of the logistic map with respect to x is r * (1 - 2x)
    const derivative = Math.abs(r * (1 - 2 * x));
    // The orbit can land exactly on the critical point x = 1/2, where the
    // derivative vanishes and log(0) = -Infinity would poison the average.
    // Skip that sample rather than returning -Infinity for the whole exponent.
    if (derivative > 0) {
      sum += Math.log(derivative);
      counted++;
    }
    x = calculateLogisticIteration(x, r);
  }

  if (counted === 0) {
    return -Infinity;
  }
  return sum / counted;
}
